Task_1_2_2

andrey-dru-me1 opened the pull request.

andrey-dru-me1 committed:
Init commit (rebase version)

andrey-dru-me1 committed:
First draft

andrey-dru-me1 committed:
add(), remove() and toString() operations have been added.

andrey-dru-me1 committed:
depthIterate has made

andrey-dru-me1 committed:
breadthIterate has made

andrey-dru-me1 committed:
build.gradle has added

andrey-dru-me1 committed:
A little test has been added

andrey-dru-me1 committed:
Javadocs have been added.

andrey-dru-me1 committed:
Beginning of Collection implementation realisation.

andrey-dru-me1 committed:
Package paths are added;
Start doing realisation of implementation Collection;
Children array are changed to a ArrayList.

andrey-dru-me1 committed:
BFS and DFS iterators are added.

andrey-dru-me1 committed:
All the implemented methods are created.
(Small issue fix)

andrey-dru-me1 committed:
nAdd() method that satisfies the task's conditions are added.

andrey-dru-me1 committed:
clear() method are fixed: now it removes the whole tree including a root.

andrey-dru-me1 committed:
Some issues are fixed and test is created.

MihailCCfit commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,251 @@
+ */
+public class Tree<T> implements Collection<T> {
+
+    private final Node root;
```
Переименуй на заглавные буквы, т.к. final, на ROOT, например

andrey-dru-me1 committed:
root in uppercase now

andrey-dru-me1 committed:
More tests were added;
Fixed some issues in the code.

andrey-dru-me1 committed:
Test coverage are increased;
Some issues fixed.

andrey-dru-me1 committed:
Huge remake of previous versions:
-Node class is private now;
-There is an id for each node instead of pointer to the Node object;
-A lot of methods are changed to work with ids;
-set and get were placed in a Tree class (not a Node);
-setById, getId are added;
-When user queries a non-existent node getNode throws IndexOutOfBoundException;
-Tests were changed to satisfy new rules.

andrey-dru-me1 committed:
Javadocs are added

andrey-dru-me1 committed:
Test has more coverage now

andrey-dru-me1 committed:
Small fix of program logic

andrey-dru-me1 committed:
Fixes pleasing the reviewdog

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+        /**
+         * Value that user places in the tree.
+         */
+        Object value;
```
А почему было не сделать `T value`?

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/358ae73044497c25c183faef56005ebc9113581a

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+                return null;   //check if current node has already removed
+            }
+
+            StringBuilder res = new StringBuilder("( " + value + ", [");
```
В актуальных версиях Java не обязательно звать билдер, стандартная конкатенация строк неплохо оптимизируется компилятором

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/91431df1ad836791eb8cc8c54b1f48a8d1f19c22

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     * @param node  Node to modify
+     * @param value Value to which the node has to be modified
+     */
+    public void set(Object node, T value) {
```
`Node node`? `T nodeValue, T newValue`?

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/395db7fd6a5f6326535f4a5584a0f19ded662cdb

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+        Object value;
+        List<Node> children;
+        Node parent;
+        private final int ID;
```
Здесь капс не нужен, это же не статическая константа

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/b262ef42b6b733f36c003d6cf6b697c4b03a2d06

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     * @return A depth-search Iterator over this tree.
+     */
+    @SuppressWarnings("unchecked")
+    public Iterator<T> iteratorDFS() {
```
То, что ты заранее перекладываешь все узлы в массив, выглядит довольно костыльно. Предполагается, что итератор работает с самим деревом, без дополнительного выделения памяти. Тебе нужно именно организовать обход по дереву в процессе работы итератора

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/9cb4e822a7cd2244ac49ee8a46346b565da76d4c

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     * @throws IndexOutOfBoundsException Throws when there isn't any
+     *                                   node with the specified identifier.
+     */
+    private Node getNodeById(int id) throws IndexOutOfBoundsException {
```
Т. е. для поиска **одного** узла ты конвертируешь **все** дерево в список? Это дикий оверхед, надо реализовать обход без конвертации

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/9cb4e822a7cd2244ac49ee8a46346b565da76d4c

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     *                                   node with specified value.
+     */
+    private Node getNode(Object value) throws IndexOutOfBoundsException {
+        for (Node i : this.toNodeList()) {
```
Все итерирования по дереву ты можешь делать просто за счет того, что оно у тебя поддерживает итератор. Т. е. остается только корректно реализовать сам итератор, а не генерить массивы на каждом шагу :)

Для всех остальных мест, где ты итерируешься по списку вместо дерева, также актуально

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/9cb4e822a7cd2244ac49ee8a46346b565da76d4c

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     * @param o Object to add
+     * @return Identifier of a new node
+     */
+    public int add2(T o) {
```
Хочется чуть более осмысленно название

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/15faf61178490fe0db09344c4222a878d2984151

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,480 @@
+     * @param o Value of node to remove
+     * @return Identifier of the removed node
+     */
+    public int remove2(Object o) {
```
Аналогично

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/15faf61178490fe0db09344c4222a878d2984151

andrey-dru-me1 committed:
Type of value is changed from Object to T

andrey-dru-me1 committed:
toString method is modified: now String is used instead of StringBuilder.

andrey-dru-me1 committed:
set method is modified

andrey-dru-me1 committed:
root and id variables are lowercased

andrey-dru-me1 committed:
Iterators updated: there isn't toNodeList() method now.

andrey-dru-me1 committed:
checkForComodification() and test for it are added.

andrey-dru-me1 committed:
add2() and remove2() are renamed to addWithId() and removeWithId().

andrey-dru-me1 committed:
Some javadocs are added.

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,573 @@
+     *
+     * @param <E> Node or T
+     */
+    private abstract class Itr<E> implements Iterator<E> {
```
Хочется более осмысленное названиие (Можно, например, `MyIterator` или `TreeIterator`)

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/e8db108565375250c836e081a9ed94ff691cc60e

molo4evan commented on Task_1_2_2/src/main/java/ru/nsu/fit/melnikov/task_1_2_2/Tree.java:
```diff
@@ -0,0 +1,573 @@
+         */
+        private final int expectedModCount;
+
+        protected final List<Node> Q;
```
На мой взгляд, удобнее вместо списка использовать Deque

andrey-dru-me1 replied:
Исправлено https://github.com/andrey-dru-me1/oop/pull/3/commits/239477f5ce4859fd376ac4836f890f91aa272601

andrey-dru-me1 committed:
More meaningful name for abstract iterator over a tree.

andrey-dru-me1 committed:
List in iterator is changed to deque.
